packages/controllers/src/snaps/SnapController.ts

```diff
@@ -1421,15 +1441,31 @@ export class SnapController extends BaseController<
     const runtime = this._getSnapRuntimeData(snapId);
     if (!runtime.installPromise) {
       console.info(`Adding snap: ${snapId}`);
-      runtime.installPromise = this._set(args as ValidatedAddSnapArgs);
+
+      // If fetching and setting the snap succeeds, this property will be set
+      // to null in the authorize() method.
+      runtime.installPromise = (async () => {
+        if ('manifest' in args && 'sourceCode' in args) {
+          return this._set({ ...args, id: snapId });
+        }
+
+        const fetchedSnap = await this._fetchSnap(snapId, args.versionRange);
+
+        return this._set({
+          ...args,
+          ...fetchedSnap,
+          id: snapId,
+        });
+      })();
```

Comment:
I pulled out the conditional `_fetchSnap` call from `_set` into this instantly invoked async arrow function, because conditionally fetching stuff from the network in a method called `_set` is cursed. (Pretty sure originally my work.) This change makes the implementation in #597 a lot more sane.